bftengine/src/bcstatetransfer/SourceSelector.cpp

```diff
@@ -175,7 +182,46 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
   fetchingTimeStamp_ = 0;
   fetchRetransmissionCounter_ = 0;
   receivedValidBlockFromSrc_ = false;
+
+  LOG_INFO(logger_, "Selected new source replica " << currentReplica_);
 }
+
+void SourceSelector::updateCurrentPrimary(uint16_t newPrimary) {
+  if (currentPrimary_ == newPrimary) return;
+  ConcordAssert(isValid(newPrimary));
+  prePrepareMsgCounter_++;
+
+  auto clearCounters = [&]() {
+    nominatedPrimary_ = NO_REPLICA;
+    prePrepareMsgCounter_ = 0;
+  };
+
+  if (nominatedPrimary_ == NO_REPLICA) {
+    nominatedPrimary_ = newPrimary;
+    return;
+  }
+
+  if (prePrepareMsgCounter_ < MAX_PREPREPARE_MSGS) {
```

Comment:
This block looks wrong in some cases, but I'm not running a unittest infront of me, so maybe i'm missing something :)

Lets see an example:
MAX_PREPREPARE_MSGS = 3
initial values nominatedPrimary_=NO_REPLICA, and currentPrimary_=0, prePrepareMsgCounter_=0
We have updateCurrentPrimary called multiple times with newPrimary=1,1,2 due to a view change and issues in the network (network partition):
1st call newPrimary with 1: prePrepareMsgCounter_=1, nominatedPrimary_=1 and exit in L201
2nd call with 1:  prePrepareMsgCounter_=2, exit on L206
3rd call with  2: prePrepareMsgCounter_=3, it jumps over block in L204 and didnt reset the counters! and chooses 2 as  currentPrimary_. This is wrong since 2 arrived only once.

If I was right, I suggest doing extensive checks in the source selector unittest, you have all the simple tools to easily write and test the code to handle all corner cases.